.github/scripts/Release.py

```diff
@@ -0,0 +1,580 @@
+        #     zipball_stream.raise_for_status()
+        #     uploaded_file = zenodo.upload_file(bucket_id=bucket_id,
+        #                                        file=zipball_stream,
+        #                                        filename=archive_filename)
+        zipball_download = requests.get(zipball_url, stream=True)
+        with tempfile.TemporaryFile() as open_tmp_file:
+            for chunk in zipball_download.iter_content():
+                open_tmp_file.write(chunk)
+            open_tmp_file.seek(0)
+            uploaded_file = zenodo.upload_file(bucket_id=bucket_id,
+                                               file=open_tmp_file,
+                                               filename=archive_filename)
+        # with open('/Users/nlf/Projects/spectre/develop/Metadata.yaml',
+        #           'rb') as open_tmp_file:
+        #     uploaded_file = zenodo.upload_file(bucket_id=bucket_id,
+        #                                        file=open_tmp_file,
+        #                                        filename=archive_filename)
+        logger.debug("Release archive upload complete:\n{}".format(
+            yaml.safe_dump(uploaded_file, allow_unicode=True)))
+
+    # Publish!
+    if auto_publish:
+        if check_only:
+            report_check_only(
+                "Would publish Zenodo record {} now!".format(new_version_id))
+        else:
+            published_record = zenodo.publish(id=new_version_id)
+            logger.debug("Zenodo record published:\n{}".format(
+                yaml.safe_dump(published_record, allow_unicode=True)))
+            logger.info(("Zenodo record is now public! Here's the link to the "
+                         "record:\n{}").format(
+                             published_record['links']['record_html']))
+    else:
+        logger.info(
+            ("Release is ready to be published on Zenodo. Go to this "
+             "website, make sure everything looks fine and then hit the "
+             "'Publish' button:\n{}").format(
+                 new_version_draft['links']['html']))
+
+
+if __name__ == "__main__":
+    # Always work with the repository that contains this file
+    repo = git.Repo(__file__, search_parent_directories=True)
+
+    import argparse
+    parser = argparse.ArgumentParser(description=(
+        "Prepare the repository and publish releases on Zenodo as part of the "
+        "automatic versioning procedure. This script is not intended to be run "
+        "outside of GitHub actions. The 'prepare' subprogram reserves a "
+        "DOI on Zenodo and inserts it into the repository along with the new "
+        "version name. Once the release archive has been created, the 'publish'"
+        "subprogram uploads it to Zenodo. Repository: {}."
+    ).format(repo.working_dir))
+    parent_parser = argparse.ArgumentParser(add_help=False)
+    parent_parser.add_argument(
+        '--zenodo-token',
```

Comment:
Question: is this the recommended way of getting tokens into scripts in CI? Mike had suggested retrieving the env variable internally to accidental printing of the API token.

Reply:
I think this is an OK way of passing secrets. Env vars also work. It's probably best to support both, though maybe not necessary for this script that's only called from the GitHub Actions (it's even in `.github/`). Side note: GitHub redacts secrets that are printed in logs, replacing them with `***`.

Reply:
I guess if we used an env var then it wouldn't need to get redacted and so in the odd case that there is a redaction bug we wouldn't leak? [not a change request, just asking/discussing for knowledge :) ]